bpf/protocol_http.h

```diff
@@ -34,6 +34,110 @@ static __always_inline http_info_t *empty_http_info() {
     return value;
 }
 
+static __always_inline u32 trace_type_from_meta(http_connection_metadata_t *meta) {
+    if (meta->type == EVENT_HTTP_CLIENT) {
+        return TRACE_TYPE_CLIENT;
+    }
+
+    return TRACE_TYPE_SERVER;
+}
+
+static __always_inline void http_get_or_create_trace_info(http_connection_metadata_t *meta,
+                                                          u32 pid,
+                                                          connection_info_t *conn,
+                                                          void *u_buf,
+                                                          int bytes_len,
+                                                          s32 capture_header_buffer) {
+    tp_info_pid_t *tp_p = tp_buf();
+
+    if (!tp_p) {
+        return;
+    }
+
+    tp_p->tp.ts = bpf_ktime_get_ns();
+    tp_p->tp.flags = 1;
+    tp_p->valid = 1;
+    tp_p->pid = pid; // used for avoiding finding stale server requests with client port reuse
+    tp_p->req_type = (meta) ? meta->type : 0;
+
+    urand_bytes(tp_p->tp.span_id, SPAN_ID_SIZE_BYTES);
+
+    u8 found_tp = 0;
+
+    if (meta) {
+        if (meta->type == EVENT_HTTP_CLIENT) {
+            pid_connection_info_t p_conn = {.pid = pid};
+            __builtin_memcpy(&p_conn.conn, conn, sizeof(connection_info_t));
+            found_tp = find_trace_for_client_request(&p_conn, &tp_p->tp);
+        } else {
+            //bpf_dbg_printk("Looking up existing trace for connection");
+            //dbg_print_http_connection_info(conn);
+
+            // For server requests, we first look for TCP info (setup by TC ingress) and then we fall back to black-box info.
+            found_tp = find_trace_for_server_request(conn, &tp_p->tp);
+        }
+    }
+
+    if (!found_tp) {
+        bpf_dbg_printk("Generating new traceparent id");
+        new_trace_id(&tp_p->tp);
+        __builtin_memset(tp_p->tp.parent_id, 0, sizeof(tp_p->tp.span_id));
+    } else {
+        bpf_dbg_printk("Using old traceparent id");
+    }
+
+    //unsigned char tp_buf[TP_MAX_VAL_LENGTH];
+    //make_tp_string(tp_buf, &tp_p->tp);
+    //bpf_dbg_printk("tp: %s", tp_buf);
+
+#ifdef BPF_TRACEPARENT
```

Comment:
Since we are moving this around, may I suggest the following?

```
// at the top of this file or some common header

#ifdef BPF_TRACEPARENT
enum { PARSE_TRACEPARENT = 1 };
#else
enum { PARSE_TRACEPARENT = 0 };
#endif 

// then here

if (PARSE_TRACEPARENT) { ... }
```

the rationale is that this ensures this code is _always_ sieved through the compiler, thus allowing it to catch any compilation error even when the macro is used. The code will _not_ be present in the final binary when PARSE_TRACEPARENT is 0 because the compiler will optimise it out anyway, so best of two worlds and less error-prone.

Another unrelated to the above thought: should this happen before the `find_trace_for_xxx(...)` code? This snippet seems to run irrespective of `found_tp`, so perhaps moving it before and bailing earlier if tp is found makes sense?

Or would it even make sense to only run this if `found_tp == false`, or is this some sort of unconditional fallback?

Reply:
I don't think the compilation is an issue, since we built all versions when compiling. Also the found_tp is not considered, because if we find something encoded in headers we don't care if we found something else before.